src/main/java/net/glowstone/entity/objects/GlowPainting.java

```diff
@@ -0,0 +1,334 @@
+        BlockFace rightFace = getRightFace();
+        double modX = rightFace.getModX() * art.getBlockWidth() / 2.0;
+        double modY = art.getBlockHeight() / 2.0;
+        double modZ = rightFace.getModZ() * art.getBlockWidth() / 2.0;
+
+        BlockFace facing = getFacing();
+        if (modX == 0.0) {
+            modX = 0.5 - facing.getModX() * 0.5 - PAINTING_DEPTH / 2;
+        } else if (modZ == 0.0) {
+            modZ = 0.5 - facing.getModZ() * 0.5 - PAINTING_DEPTH / 2;
+        }
+
+        Location add = getTopLeftCorner().add(modX, -modY, modZ);
+        location.setX(add.getX());
+        location.setY(add.getY());
+        location.setZ(add.getZ());
+        location.setPitch(0);
+        location.setYaw(getYaw());
+    }
+
+    @Override
+    public boolean setFacingDirection(BlockFace blockFace, boolean force) {
+        HangingFace oldFace = facing;
+        this.facing = HangingFace.getByBlockFace(blockFace);
+
+        if (!force && isObstructed()) {
+            this.facing = oldFace;
+            return false;
+        }
+
+        respawn();
+
+        return true;
+    }
+
+    @Override
+    public void setFacingDirection(BlockFace blockFace) {
+        setFacingDirection(blockFace, false);
+    }
+
+    @Override
+    public void pulse() {
+        super.pulse();
+
+        if (ticksLived % 11 == 0) {
+
+            if (location.getBlock().getRelative(getAttachedFace()).getType() == Material.AIR) {
+                if (EventFactory.callEvent(new HangingBreakEvent(this, RemoveCause.PHYSICS)).isCancelled()) {
+                    return;
+                }
+                world.dropItemNaturally(location, new ItemStack(Material.PAINTING));
+                remove();
+            }
+        }
+
+        if (ticksLived % (20 * 5) == 0 && isObstructed()) {
+            if (EventFactory.callEvent(new HangingBreakEvent(this, RemoveCause.PHYSICS)).isCancelled()) {
+                return;
+            }
+            if (location.getBlock().getRelative(getAttachedFace()).getType() == Material.AIR) {
+                world.dropItemNaturally(location, new ItemStack(Material.PAINTING));
+                remove();
+            }
+        }
+    }
+
+    @Override
+    protected void pulsePhysics() {
+        // not affected by physics
+    }
+
+    public boolean isObstructed() {
+        Location current = getTopLeftCorner();
+        BlockFace right = getRightFace();
+
+        for (int y = 0; y < art.getBlockHeight(); y++) {
+            for (int x = 0; x < art.getBlockWidth(); x++) {
+                if (!canHoldPainting(current)) {
+                    return true;
+                }
+                current = current.getBlock().getRelative(right).getLocation();
+            }
+
+            current = current.getBlock().getRelative(BlockFace.DOWN).getLocation();
+
+            // reset x and z
+            current.subtract(right.getModX() * art.getBlockWidth(), 0, right.getModZ() * art.getBlockWidth());
```

Comment:
Since you are using `substract`, make sure the locations are cloned before updating them, IIRC it can update the block's location too (unless that only happens on Spigot?)